core/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -238,4 +269,49 @@ public void messageReceived(NodeRequest request, TransportChannel channel) throw
         }
 
     }
+
+    /**
+     * Contains the responses (success and failure) for multi-node requests.
+     * <p>
+     * The expectation is that the responses contained by it will map to every node with a single response (success <em>or</em> failure)
+     * per node.
+     */
+    public static class ResponseContainer<T extends BaseNodeResponse> {
+
+        private final T[] nodeResponses;
+        private final FailedNodeException[] failures;
+
+        /**
+         * Create the {@code ResponseContainer}.
+         *
+         * @param nodeResponses The list of expected responses
+         * @param failures The list of node failures
+         * @throws NullPointerException if any parameter is {@code null}
+         */
+        @SuppressWarnings("unchecked")
+        public ResponseContainer(Class<T> clazz, List<T> nodeResponses, List<FailedNodeException> failures) {
+            this.nodeResponses = nodeResponses.toArray((T[])Array.newInstance(clazz, nodeResponses.size()));
```

Comment:
> I'd consider converting everything in `BaseNodesResponse` to a list instead of an array and then just passing the list around?

@pickypg Can we do this?